### ISSUE TITLE
Extension properties setters does not work as expected

### ISSUE THREAD

**Issue by ibaklan** [lampepfl/dotty#5588]
After recently merged [extension methods pull request](https://github.com/lampepfl/dotty/pull/5114) extension properties definition (getters and setters) become available.
Actually it is possible to define extension setter (together with extension getter), but on use-site only extension getter works as expected, attempt to use extension setter leads to compilation error.

Demonstrating examples (extension activation via implicits or via import - same result)
Example 1 (using import)
```Scala
object TestMain {
  def main(args: Array[String]): Unit = {
    testExtensionProperty()
  }

  case class Foo(var foo: String)

  object fooExt {
    // define `Foo`-extension property with getter and setter delegating to `Foo.foo`
    def (thisFoo: Foo) fooExt: String = thisFoo.foo
    def (thisFoo: Foo) fooExt_= (value: String): Unit = { thisFoo.foo = value }
  }

  def testExtensionProperty(): Unit = {
    import fooExt._
    val foo = Foo("initVal")
    assert(foo.fooExt == "initVal")
    foo.fooExt = "updatedVal"
    assert(foo.foo == "updatedVal")
    assert(foo.fooExt == "updatedVal")
  }
}
```

Example 2 (using implicits)
```Scala
object TestMain2 {
  def main(args: Array[String]): Unit = {
    testExtensionProperty()
  }

  case class Foo(var foo: String)

  implicit object fooExt {
    // define `Foo`-extension property with getter and setter delegating to `Foo.foo`
    def (thisFoo: Foo) fooExt: String = thisFoo.foo
    def (thisFoo: Foo) fooExt_= (value: String): Unit = { thisFoo.foo = value }
  }

  def testExtensionProperty(): Unit = {
    //import fooExt._
    val foo = Foo("initVal")
    assert(foo.fooExt == "initVal")
    foo.fooExt = "updatedVal"
    assert(foo.foo == "updatedVal")
    assert(foo.fooExt == "updatedVal")
  }
}
```

**Expected behaviour** - code snippets should be compilable, asserts should be passed successfully

**Actual behavior** - code compilation fails with error
```
    foo.fooExt = "updatedVal"
    ^^^^^^^^^^^^^^^^^^^^^^^^^
    reassignment to val `fooExt`
```

**FYI**
[Kotlin extension properties](https://kotlinlang.org/docs/reference/extensions.html#extension-properties) works as expected (with both getters and setters):
Same (working) example in Kotlin may look like this:
```Kotlin
object TestMain {

    @JvmStatic
    fun main(arg: Array<String>) {
        testExtensionProperty()
    }

    data class Foo(var foo: String)

    var Foo.fooExt : String
        get() = foo
        set(value: String) { foo = value }

    fun testExtensionProperty() {
        val foo = Foo("initVal")
        assert(foo.fooExt == "initVal")
        foo.fooExt = "updatedVal"
        assert(foo.foo == "updatedVal")
        assert(foo.fooExt == "updatedVal")
    }
}
```

**Comment by odersky** [comment]
According to the current spec, `p.x = y` is translated to `p.x_=(y)` if the prefix `p` has a member `x_=`. That's not the case for extension methods. 

So this would require a spec change, and implementing it would be a considerable complication of the compiler. I am not sure it's worth it, honestly. What's the point to have extension method setters? In particular if you want to encourage a functional style?

**Comment by ibaklan** [comment]
Well, my thought was that extension method should be as much "indistinguishable" to regular methods as it ever possible. At least this idea looks quite natural for me.
This particular issue was trigger by [this post](https://contributors.scala-lang.org/t/pre-sip-thisfunction-scope-injection-similar-to-kotlin-receiver-function/2234/41) (and following discussion) - that just reminds me to check whether extension setters works or not.

But in general I think that all that: 
* property setters(`name_=`), `apply`, `update`, 
* transparent access to implicit members of `this` (extension target),
* access with or without `this.` prefix 

**-** all that features of regular member methods could be "expect-able" for extension methods.
If all of those before mentioned features would be available with extension methods as well then it could be said that there are no gaps remain in extension methods specification.

Regarding this particular case (extension property with setters), one may admit,  probably there are not so much use case:
1. define custom strictly typed properties on some "HasMap"-like object <br/>(like `foo.barExt` <==> `foo.map(BAR_KEY):BarType`)
2. defining transformed properties <br/> like `foo.barExt` <==> `Transformation(foo.bar)`
3. extension property could be some alias to long path, like `foo.barBaxExt` == `foo.bar.baz`

No other use cases I could imagine for now. More interesting could be use case with indirection (highlighted later)

But overall, I would say that major motivation for supporting extension properties could be "political".
Comparing Scala and Kotlin extensions methods (and other "extensions related stuff") one may come to conclusion that Kotlin's implementation has less gaps.

In this case willing to close/fill that gaps in Scala extensions "stuff" implementation may send clearly good signal about Scala bright future (for all who may perform that Scala vs Kotlin comparision).

And vise versa - willing to not fill that gabs may send bad signal for all who consider different language alternatives (sometimes people may make their decisions basing on minor things treating them as signs of overall altitude).

From this perspective filling even minor gaps may have good effect (especially when that gaps are already filled in competitors implementation). However this is only my personal feelings - it might be not objective (or wrong).

**Comment by ibaklan** [comment]
From the other perspective Kotlin's extensions-related stuff is much close to [C++'s Pointers to members](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.cbclx01/cplr034.htm)
With the only difference in how does access to them look like
* in C++ it looks like `targetObj.*memberPointer()` 
* while in Kotlin it will be just `targetObj.memberPointer()` assuming definitions like <br/> `val targetObj: Foo = ...` and `var memberPointer: Foo.() -> Bar = ...`

But what is also missing in Kotlin - is something similar to C++'s pointer to fields (not only to methods).

This feature (I guess) could be even more useful then regular extensions properties (effectively it should be closer to **extensions functions type**, but those **"extensions assignable functions (properties?) types"** should be something like regular **extension functions types**, also equipped with `update` implementation along with `apply` one).

Although (any way) it is quite separate feature.

**Comment by ibaklan** [comment]
**FYI:** comparison of **Kotlin's extensions related features** vs **Scala's extensions related features** could be summarized in following table:

| Feature | Kotlin | Scala |
| ---     | ---    | ---   |
| extension methods <br/> members and local | **yes** <br/> [doc](https://kotlinlang.org/docs/reference/extensions.html) <br/> (both members and locals are supported) | **yes** <br/> (both members and locals are supported)
| extension function types <br/>  (`val`s / `var`s which defines indirect extension methods) | **yes** <br/> [doc](https://kotlinlang.org/docs/reference/lambdas.html#function-types) , [doc](https://kotlinlang.org/docs/reference/lambdas.html#function-literals-with-receiver) | **no**  <br/> there is no way to define <br/> `val barExt: SomeExtType[Foo] = ...` and then use it immediately as `foo.barExt()`
| extension lambdas <br/> (lambdas with "overloaded" meaning of this) | **yes** <br/> [doc](https://kotlinlang.org/docs/reference/lambdas.html#function-literals-with-receiver), [post](https://stackoverflow.com/questions/48558252/kotlin-extension-lambdas-vs-regular-lambdas) (`this` in extension lambda body points to first parameter, since `this.` prefix is optional it leads to type safe scope injection) | **???** <br/> implicit functions lambdas could be used, but scope injection occurs not into regular scope but into implicits scope <br/> [discussion1](https://contributors.scala-lang.org/t/pre-sip-thisfunction-scope-injection-similar-to-kotlin-receiver-function/2234), [discussion2](https://contributors.scala-lang.org/t/implicit-function-types/219/39) ... |
| type safe dsl definition | **yes** <br/> [doc: Type-Safe Builders](https://kotlinlang.org/docs/reference/type-safe-builders.html) - build on top of "extension lambdas" / "extension function types" | **yes** <br/> but with some limitations: <br/>(&#42;)  implementation based on implicit functions types suffers "global scope pollution", which also lead to some [issues](https://contributors.scala-lang.org/t/delayedinit-or-oncreate-any-solution/1748/29) <br/> (&#42;) [possible workaround](https://contributors.scala-lang.org/t/pre-sip-thisfunction-scope-injection-similar-to-kotlin-receiver-function/2234/47) is too tricky
| extension properties | **yes** <br/> [doc](https://kotlinlang.org/docs/reference/extensions.html#extension-properties) - both getters and setters supported |  **yes/no** <br> (only getters are supported)
| `this.` prefix is optional when <br/> accessing extensions applicable to `this` | **yes** <br/> this feature heavily used with extension function types <br/>see [apply function impl](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/src/kotlin/util/Standard.kt#L75) as example: there `block` is "indirect extension method" (`val` of ext. function type), `block()` invocation is in fact resolves to `this.block()` invocation | **no**

**Comment by odersky** [comment]
Fixed by #9552
